In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

In [9]:
df = pd.read_excel("c:\\Users\\shaba\\PycharmProjects\\Data-Analysis\\Data\Online Retail.xlsx")

<>:1: SyntaxWarning: invalid escape sequence '\O'
<>:1: SyntaxWarning: invalid escape sequence '\O'
C:\Users\shaba\AppData\Local\Temp\ipykernel_12204\3107174252.py:1: SyntaxWarning: invalid escape sequence '\O'
  df = pd.read_excel("c:\\Users\\shaba\\PycharmProjects\\Data-Analysis\\Data\Online Retail.xlsx")


In [10]:
df.sample(5)

InvoiceNo StockCode                   Description  Quantity  \
117917    546414     22649  STRAWBERRY FAIRY CAKE TEAPOT         2   
446747    574918     84836   ZINC METAL HEART DECORATION        24   
103816    545078     22099     CARAVAN SQUARE TISSUE BOX         2   
221254    556243     22355     CHARLOTTE BAG SUKI DESIGN         2   
230696    557214     22726    ALARM CLOCK BAKELIKE GREEN         4   

               InvoiceDate  UnitPrice  CustomerID         Country  
117917 2011-03-13 10:25:00       4.95     16768.0  United Kingdom  
446747 2011-11-07 16:31:00       1.25         NaN          France  
103816 2011-02-28 10:15:00       2.46         NaN  United Kingdom  
221254 2011-06-09 16:07:00       1.63         NaN  United Kingdom  
230696 2011-06-17 12:20:00       3.75     17677.0  United Kingdom

In [11]:
df.shape

(541909, 8)

In [17]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [21]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [25]:
df.duplicated().sum()

np.int64(5268)

In [26]:
df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  541909.000000                         541909  541909.000000   
mean        9.552250  2011-07-04 13:34:57.156386048       4.611114   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 11:34:00       1.250000   
50%         3.000000            2011-07-19 17:17:00       2.080000   
75%        10.000000            2011-10-19 11:27:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       218.081158                            NaN      96.759853   

          CustomerID  
count  406829.000000  
mean    15287.690570  
min     12346.000000  
25%     13953.000000  
50%     15152.000000  
75%     16791.000000  
max     18287.000000  
std      1713.600303

## Quality
- Missing values in Description and CustomerID
- CustomerID datatype is float and InvoiceDate datatye is datetime64[ns]
- Negative values in quantity
- inconsistency in InvoiceNo and StockCode i.e. contains alphabets at start in some InvoiceNo and at end In some StockCode 

## Tidiness

- Same column for date and time i.e. InvoiceDate
- 5268 duplicate rows

In [43]:
data = df.copy()

In [58]:
g = data.groupby('CustomerID')
